File: python/binary_tree_zigzag_level_order_traversal.py
#!/usr/bin/env python

# Definition for a  binary tree node
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    # @param root, a tree node
    # @return a list of lists of integers
    def zigzagLevelOrder(self, root):
        if not root:
            return []
        l2r = True
        result, queue = [], collections.deque([root])
        while queue:
            level, size = [], len(queue)
            for i in xrange(size):
                node = queue.popleft()
                if node.left:
                    queue.append(node.left)
                if node.right:
                    queue.append(node.right)
                level.append(node.val)
            if l2r:
                result.append(level)
            else:
                result.append(level[::-1])
            l2r = not l2r
        return result

File: java/BinaryTreeZigzagLevelOrderTraversal.java
import java.util.Collections;
import java.util.LinkedList;
import java.util.List;
import java.util.Queue;

/**
 * Created by Robbert on 1/30/15.
 * Binary Tree Zigzag Level Order Traversal
 * https://oj.leetcode.com/problems/binary-tree-zigzag-level-order-traversal/
 */
public class BinaryTreeZigzagLevelOrderTraversal {
    public List<List<Integer>> zigzagLevelOrder(TreeNode root) {
        List<List<Integer>> result = new LinkedList<List<Integer>>();
        if (root == null) {
            return result;
        }

        boolean l2r = true;
        Queue<TreeNode> queue = new LinkedList<TreeNode>();
        queue.offer(root);
        while (!queue.isEmpty()) {
            int size = queue.size();
            List<Integer> level = new LinkedList<Integer>();
            for (int i = 0; i < size; i++) {
                TreeNode node = queue.poll();
                if (node.left != null) {
                    queue.offer(node.left);
                }
                if(node.right != null) {
                    queue.offer(node.right);
                }
                level.add(node.val);
            }
            if (!l2r) {
                Collections.reverse(level);
            }
            l2r = !l2r;
            result.add(level);
        }

        return result;
    }
}